File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>

using namespace std;

int n=0;
int arr[8][2];
int visited[8];
int answer = 0;

void bt(int d, int x, int cnt){
    
    answer = max(answer,cnt);
    for(int i=0;i<n;i++){
        if(visited[i]==1) continue;
        if(x >= arr[i][0]){
            visited[i]=1;
            bt(i+1, x-arr[i][1],cnt+1);
            visited[i]=0;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    n = dungeons.size();
    
    for(int i=0;i<dungeons.size();i++){
        for(int j=0;j<dungeons[i].size();j++){
            arr[i][j] = dungeons[i][j];
        }
    }
    bt(0,k,0);
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    
    private static int answer = 0;
    private static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length+1];
        
        bt(k,0,dungeons);
        
        return answer;
    }
    
    private static void bt(int Hp, int cnt, int[][]dg){
        
        for(int i=0;i<dg.length;i++){
            if(!visited[i] && dg[i][0] <=Hp){
                visited[i] = true;
                bt(Hp - dg[i][1], cnt+1, dg);
                visited[i] = false;
            }
        }
        answer = Math.max(answer, cnt);
    }
}